### ISSUE TITLE
Backward capability roadmap — recompute, dual loops, softmax/LN/Welford/GEMM gradients

### ISSUE THREAD

**Issue by Lizn-zn** [Lizn-zn/VeriTile#13]
**Tracking consolidated on 2026-09-24 into #106.**

Remaining scope: Future backward breadth: audit what existing FA-1/FA-2 proofs actually cover, then select concrete remaining softmax/norm/GEMM gradient cases. AD-based generation remains a separate research option.

This detail issue is closed as **not planned separately** to keep one active topic entry. The work is not declared implemented or abandoned; continue tracking it in #106, which retains this acceptance scope and links back here. Reopen a focused issue when it becomes an active implementation tranche. The original request and discussion are preserved below.

---

## Scope

Capability roadmap for **backward kernel verification** across the kernel families VeriTile targets in forward. Tracks the design space; concrete FA-1 backward work lives in **#43**.

This issue is the *capability index*. Per-kernel backward work should be either added as comments here when scoped, or split into focused issues like #43 was for FA-1.

## Why backward is its own technical territory (not a small extension of forward)

Forward kernel `(naive ↔ optimized)` equivalence proofs already have an invariant-driven harness. Backward introduces **at least four new challenges**:

### 1. Recompute data flow

FA-backward's IO-aware trick is to **not store** forward's `P = softmax(QK^T)` and recompute it during backward. Proving \"recomputed `P` equals forward's `P`\" requires linking forward and backward execution traces — one needs to argue that under identical inputs, the recomputed intermediate equals what forward used. VeriTile's current `exec : Kernel → BlockState → Option BlockState` interface does not directly support \"two kernels share a hypothesis\" propositions; this is an infrastructure question for backward.

### 2. Reverse-mode AD as spec

Forward spec is `attentionReal Q K V causal` (algorithm-level attention output). Backward spec is naturally the **reverse-mode AD of `attentionReal`**:

```
attentionBackwardReal Q K V dO :
  Matrix (Fin S) (Fin D) ℝ × Matrix (Fin S) (Fin D) ℝ × Matrix (Fin S) (Fin D) ℝ
  -- (dQ, dK, dV)
```

This requires defining `∂/∂Q`, `∂/∂K`, `∂/∂V` over ℝ and unfolding to per-element expressions for kernel-equivalence comparison — not a Mathlib `fderiv` reference, since downstream the spec must be compared elementwise to kernel output.

### 3. Softmax gradient algebraic identity

Central algebra:

```
dS_ij = P_ij * (dP_ij - Σ_k P_ik * dP_ik)
```

(`P` is attention probability, `dP = dO · V^T`.) This \"fused softmax JVP\" is the central backward algebraic lemma. Comparable in difficulty to the online-softmax recurrence — forward is a single-variable streaming recurrence, JVP is reverse-reduce + elementwise correction; `simp` normal form design has to be redone from scratch.

### 4. dQ atomic dependency (some implementations)

Some FA-backward implementations (FA-1 backward `dQ` accumulation in particular) use **`atomic_add`** across K-blocks because `dQ`'s contributions from multiple K-blocks reduce to the same row. This **directly depends on #12 atomic / async semantics** — without #12, production FA-1 backward is not reachable; only stripped versions (sequential reduce or rearranged loops to avoid atomic) are.

## Kernel families covered

| Forward | Backward shape |
|---|---|
| Softmax | softmax JVP: `dx_i = y_i * (dy_i - Σⱼ y_j * dy_j)` |
| LayerNorm / RMSNorm | dual gradient (w.r.t. `γ`, `β`, w.r.t. input `x`); second-moment chain rule |
| Welford | online-variance recurrence backward; `(M, S, n)` chain rule unfolding |
| FlashAttention forward | **FA-backward** (Dao 2022 / 2023): recompute `P`, dual outer loops (Q-block computes `dK, dV`, K-block computes `dQ`) — see **#43** for FA-1-specific work |
| GEMM forward | GEMM backward triple: `dA = dC · B^T`, `dB = A^T · dC`, bias gradient |
| Element-wise activation | chain rule + elementwise `σ'` |

### A. Forward-only (status quo)

- Pro: zero additional work
- Con: training-time kernels fully uncovered

### B. Math-only — formalize backward specs and algebraic identities

Define `attentionBackwardReal` (and per-kernel analogues), prove softmax JVP and recurrence-backward algebraic lemmas. Math-only, no kernel side.

- Pro: small (~100–200 lines per family); independent prerequisite for any kernel-side work
- Con: not a verified-rewrite, just math
- Note: this is **the lowest-cost first step** — without backward specs defined, kernel equivalence theorems cannot even be stated

### C. Stripped backward — no atomic, single-pass

Simplified production backward: single Q-block, explicit `P` recompute, sequential reduce in place of atomic. Prove `stripped_X_backward ≡ X_backward_math`.

- Pro: doesn't depend on #12; reuses forward's invariant tooling
- Con: large gap from production (no atomic is unrealistic); `flash_attn_backward` diff-test won't bit-match

### D. Full backward — production form, atomic dQ

Complete FA-1 / FA-2 backward, directly corresponds to `flash_attn_backward`.

- Pro: full capability; can verify production training kernels
- Con: **strongly depends on #12**; cannot start before atomic semantics land

### E. Auto-diff route — derive backward from forward via syntactic AD

Instead of \"backward kernel ≡ math\", prove \"backward kernel ≡ syntactic AD of forward.\" This is the route of Coq/Lean AD formalization (Smolka et al.).

- Pro: each forward kernel automatically gets a backward equivalence; no per-kernel re-proof
- Con: syntactic AD is a research problem; Lean 4 infrastructure unclear; production backward is hand-optimized (recompute, fused JVP), so an additional \"AD output ≡ hand-optimized form\" layer would be needed
- Status: research-level; standalone effort, not on the near-term roadmap

## Dependencies

- **#12 (async / atomic)**: D requires; C explicitly avoids via sequential reduce
- **#11 (fp semantics)**: softmax JVP `P_ij * (dP_ij - Σ_k P_ik * dP_ik)` has catastrophic cancellation risk in fp when `dP_ij ≈ Σ_k P_ik * dP_ik`; #11's bounded-error route is more important for backward than for forward
- **#10 (lifter)**: lifter design is forward/backward-agnostic; backward kernel lifting is an extension after lifter ships
- **headline corollary `fa1_eq_fa2`**: does **not** extend to backward — FA-1 backward and FA-2 backward have different IO patterns (FA-1 backward: dual outer loops; FA-2 backward: single outer loop + different work partitioning); `fa1_backward ≡ fa2_backward` needs separate proof, not a corollary

## Suggested ordering (capability-side)

```
Foundational (math-only, no kernel work):
  Option B for FA-1 — see #43 Sub-2a (math spec + softmax JVP)
  Option B for softmax / LayerNorm / Welford / GEMM — analogous math specs

Kernel-side, doesn't need atomic:
  Option C for FA-1 — see #43 Sub-2b (stripped backward)
  Option C for other kernels as analogous

Kernel-side, needs atomic:
  Option D for FA-1 — see #43 Sub-2c (after #12)

Research-frontier, on hold:
  Option E — pending Lean 4 AD infrastructure
```

## Related issues

- **#43** — FA-1 backward specific (LSE forward output + dQ/dK/dV kernel + softmax JVP identity)
- **#11** — fp semantics, especially relevant to backward's cancellation-prone JVP
- **#12** — async / atomic, hard dependency for option D
- **#10** — Python lifter, extension question for backward kernels
